File: frontend/src/types/schemas.ts
// Mirror of backend/schemas.py. Field names and ranges must match exactly so
// the frontend can consume real pipeline output without translation.

export interface ParsedHypothesis {
  claim: string;
  mechanism: string;
  context: string;
  population: string;
  method: string;
}

export interface Paper {
  paper_id: string;
  title: string;
  authors: string[];
  year: number | null;
  abstract: string;
  url: string;
  citation_count: number;
  relevance_score: number;
  cluster: string;
}

export interface Conflict {
  paper_id: string;
  title: string;
  disagreement_dimension: string;
  explanation: string;
  severity: number; // 0..1
}

export interface OverlapReport {
  crowding_score: number; // 0..100
  overlapping_papers: string[];
  whitespace_summary: string;
  risk_notes: string[];
}

export interface ResearchGroup {
  group_id: string;
  name: string;
  institution: string;
  principal_investigators: string[];
  recent_paper_ids: string[];
  methods: string[];
  grounding_evidence: string;
}

export interface EmulatorOutput {
  group_id: string;
  group_name: string;
  interest_score: number; // 0..100
  engagement_type: string;
  proposed_direction: string;
  method_they_use: string;
  time_to_publish_months: number;
  competitive_risk: number; // 0..100
  grounding_paper_ids: string[];
}

export interface Scenario {
  scenario_id: string;
  name: string;
  probability: number; // 0..1
  description: string;
  leading_groups: string[];
  implications: string[];
}

export interface ImpactDimension {
  score: number; // 0..100
  confidence_low: number;
  confidence_high: number;
  rationale: string;
}

export interface ImpactForecast {
  volume: ImpactDimension;
  velocity: ImpactDimension;
  reach: ImpactDimension;
  depth: ImpactDimension;
  disruption: ImpactDimension;
  translation: ImpactDimension;
  overall_summary: string;
}

export type ImpactDimensionKey =
  | "volume"
  | "velocity"
  | "reach"
  | "depth"
  | "disruption"
  | "translation";

export const IMPACT_DIMENSIONS: ImpactDimensionKey[] = [
  "volume",
  "velocity",
  "reach",
  "depth",
  "disruption",
  "translation",
];

export interface Variant {
  variant_id: string;
  hypothesis_text: string;
  operator: string;
  rationale: string;
  impact_scores: Partial<Record<ImpactDimensionKey, number>>;
  is_pareto_selected: boolean;
  dominance_explanation: string;
}

export interface GroundednessCheck {
  group_id: string;
  group_name: string;
  is_grounded: boolean;
  flagged_inconsistencies: string[];
  evidence: string;
}

export interface StrategyMemo {
  recommendation: string;
  executive_summary: string;
  key_findings: string[];
  selected_variants: string[];
  risks: string[];
  next_steps: string[];
}

// Mirrors backend.pipeline.PipelineState. Frontend-only display extras
// (denario_outline, next_actions, headline_groups) are deliberately separate
// so they don't clash with the backend's authoritative shape.
export interface PipelineState {
  raw_hypothesis: string;
  parsed?: ParsedHypothesis;
  papers: Paper[];
  conflicts: Conflict[];
  overlaps?: OverlapReport;
  groups: ResearchGroup[];
  emulator_outputs: EmulatorOutput[];
  scenarios: Scenario[];
  forecast?: ImpactForecast;
  variants: Variant[];
  groundedness_checks: GroundednessCheck[];
  final_memo?: StrategyMemo;
}

// Frontend-only additions used for demo presentation; not produced by the backend.
export interface DenarioOutline {
  title: string;
  abstract: string;
  methods: string;
  experiments: string;
  discussion: string;
}

export interface DemoExtras {
  headline_groups: string[]; // group_ids the narrator zooms into
  denario_outline: DenarioOutline;
  next_actions: string[];
}

export interface DemoState extends PipelineState {
  extras: DemoExtras;
}

File: backend/schemas.py
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class ParsedHypothesis(BaseModel):
    claim: str
    mechanism: str
    context: str
    population: str
    method: str


class Paper(BaseModel):
    paper_id: str
    title: str
    authors: List[str] = Field(default_factory=list)
    year: Optional[int] = None
    abstract: str = ""
    url: str = ""
    citation_count: int = 0
    relevance_score: float = 0.0
    cluster: str = "unclustered"


class Conflict(BaseModel):
    paper_id: str
    title: str
    disagreement_dimension: str
    explanation: str
    severity: float = Field(ge=0.0, le=1.0)


class OverlapReport(BaseModel):
    crowding_score: int = Field(ge=0, le=100)
    overlapping_papers: List[str] = Field(default_factory=list)
    whitespace_summary: str
    risk_notes: List[str] = Field(default_factory=list)


class ResearchGroup(BaseModel):
    group_id: str
    name: str
    institution: str
    principal_investigators: List[str] = Field(default_factory=list)
    recent_paper_ids: List[str] = Field(default_factory=list)
    methods: List[str] = Field(default_factory=list)
    grounding_evidence: str


class EmulatorOutput(BaseModel):
    group_id: str
    group_name: str
    interest_score: int = Field(ge=0, le=100)
    engagement_type: str
    proposed_direction: str
    method_they_use: str
    time_to_publish_months: int = Field(ge=0)
    competitive_risk: int = Field(ge=0, le=100)
    grounding_paper_ids: List[str] = Field(default_factory=list)


class Scenario(BaseModel):
    scenario_id: str
    name: str
    probability: float = Field(ge=0.0, le=1.0)
    description: str
    leading_groups: List[str] = Field(default_factory=list)
    implications: List[str] = Field(default_factory=list)


class ImpactDimension(BaseModel):
    score: int = Field(ge=0, le=100)
    confidence_low: int = Field(ge=0, le=100)
    confidence_high: int = Field(ge=0, le=100)
    rationale: str


class ImpactForecast(BaseModel):
    volume: ImpactDimension
    velocity: ImpactDimension
    reach: ImpactDimension
    depth: ImpactDimension
    disruption: ImpactDimension
    translation: ImpactDimension
    overall_summary: str


class MetricScore(BaseModel):
    metric_name: str
    score: int = Field(ge=0, le=100)
    confidence_low: int = Field(ge=0, le=100)
    confidence_high: int = Field(ge=0, le=100)
    rationale: str
    evidence: List[str] = Field(default_factory=list)
    weakness: str = ""


class Scorecard(BaseModel):
    composite_score: int = Field(ge=0, le=100)
    verdict: str
    metric_scores: List[MetricScore] = Field(default_factory=list)
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    evidence_summary: str


class Variant(BaseModel):
    variant_id: str
    hypothesis_text: str
    operator: str
    rationale: str
    impact_scores: Dict[str, int] = Field(default_factory=dict)
    composite_score: int = Field(default=0, ge=0, le=100)
    scorecard: Optional[Scorecard] = None
    rank: Optional[int] = None
    is_pareto_selected: bool = False
    dominance_explanation: str = ""


class GroundednessCheck(BaseModel):
    group_id: str
    group_name: str
    is_grounded: bool
    flagged_inconsistencies: List[str] = Field(default_factory=list)
    evidence: str


class StrategyMemo(BaseModel):
    recommendation: str
    executive_summary: str
    key_findings: List[str] = Field(default_factory=list)
    selected_variants: List[str] = Field(default_factory=list)
    risks: List[str] = Field(default_factory=list)
    next_steps: List[str] = Field(default_factory=list)

    def to_markdown(self) -> str:
        sections = [
            "# Strategy Memo",
            f"## Recommendation\n{self.recommendation}",
            f"## Executive Summary\n{self.executive_summary}",
            _bullets("Key Findings", self.key_findings),
            _bullets("Selected Variants", self.selected_variants),
            _bullets("Risks", self.risks),
            _bullets("Next Steps", self.next_steps),
        ]
        return "\n\n".join(section for section in sections if section)


def _bullets(title: str, values: List[str]) -> str:
    if not values:
        return ""
    body = "\n".join(f"- {value}" for value in values)
    return f"## {title}\n{body}"
